### PR TITLE
feature(new integration): Adroll Web Device Mode 

### PR DESCRIPTION
**Pull request by yashasvibajpai**
## Description of the change

> Onboarding Adroll integration

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [x] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

---
This change is [Reviewable](https://reviewable.io/reviews/rudderlabs/rudder-sdk-js/504)

### CODE REVIEW

File: integrations/Adroll/browser.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+} from "../utils/commonUtils";
+
+import { NAME } from "./constants";
+import ScriptLoader from "../ScriptLoader";
+
+class Adroll {
+  constructor(config) {
+    this.advId = config.advId;
+    this.pixId = config.pixId;
+    // this.pixelId = config.pixelId;
```
Comment (utsabc) [inline integrations/Adroll/browser.js:20]: Lets remove this commented lines

File: integrations/Adroll/browser.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    logger.debug("===In isReady Adroll===");
+    return !!window.__adroll;
+  }
+
+  identify(rudderElement) {
+    logger.debug("===In Adroll Identify===");
+    const { message } = rudderElement;
+
+    let payload = {
+      adroll_email: get(message, "context.traits.email"),
```
Comment (utsabc) [inline integrations/Adroll/browser.js:45]:
```suggestion
      adroll_email: get(message, "traits.email") || get(message, "context.traits.email"),
```

File: integrations/Adroll/browser.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    };
+
+    if (!payload.adroll_email) {
+      logger.error("User parameter (email) is required for identify call");
+      return;
+    }
+    window.__adroll.record_adroll_email('segment');
```
Comment (utsabc) [inline integrations/Adroll/browser.js:52]: Can you add some adroll documentation for this statement

File: integrations/Adroll/browser.js [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+import { NAME } from "./constants";
+import ScriptLoader from "../ScriptLoader";
+
+class Adroll {
+  constructor(config) {
+    this.advId = config.advId;
+    this.pixId = config.pixId;
+    this.name = NAME;
+    window.adroll_adv_id = this.advId;
+    window.adroll_pix_id = this.pixId;
```
[the remark this answers is not in the document]
Comment (yashasvibajpai) [reply integrations/Adroll/browser.js:17]: In the native adroll script, I could see references to `adroll_adv_id` and `adroll_pix_id` only, hence used the same.

File: integrations/Adroll/browser.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+      rudderElement.message.properties,
+      "revenue"
+    );
+    if (userId) {
+      properties.user_id = userId;
+    }
+    if (properties.price) {
+      properties.adroll_conversion_value = properties.price;
+      delete properties.price;
+    }
```
[the remark this answers is not in the document]
Comment (yashasvibajpai) [reply integrations/Adroll/browser.js:68]: This was intentional as there was a difference regarding mapping price or revenue to `adroll_conversion_value` in docs and code. Removing this.

File: integrations/Adroll/browser.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (ItsSudip) [reply integrations/Adroll/browser.js:17]: It is not required I think. When you initialize the SDK that should automatically declare those variables if they are using internally. Check once if it is not the case.

File: integrations/Adroll/browser.js [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+      "order completed",
+      "order cancelled",
+      "order updated",
+    ];
+    const { message } = rudderElement;
+    const { userId, event, properties } = message;
+    const eventsHashmap = getHashFromArray(this.eventsMap);
+    let data;
+    if (eventsHashmap[event.toLowerCase()]) {
+      if (PRODUCT_EVENTS.includes(event.toLowerCase())) {
```
Comment (utsabc) [inline integrations/Adroll/browser.js:73]: Lets use indexOf to check if it exists in array for older versions of browser includes wont be supported

File: integrations/Adroll/util.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
+/* eslint-disable no-param-reassign */
+const productEvent = (properties) => {
+  if (properties.price) {
+    properties.adroll_conversion_value = properties.price;
+    delete properties.price;
+  }
+  return properties;
+};
+
+const orderEvent = (properties) => {
```
Comment (utsabc) [inline integrations/Adroll/util.js:10]: Lets add some inline documentation for this

**Review by ItsSudip — DISMISSED** [review]

File: integrations/Adroll/browser.js [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+      "product viewed",
+      "product added",
+    ];
+    const ORDER_EVENTS = [
+      "cart viewed",
+      "checkout started",
+      "order completed",
+      "order cancelled",
+      "order updated",
+    ];
```
Comment (shrouti1507) [inline integrations/Adroll/browser.js:67]: can we clean this up and move the declarations to other file and import.

File: integrations/Adroll/browser.js [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        data = productEvent(properties);
+      } else if (ORDER_EVENTS.indexOf(event.toLowerCase()) !== -1) {
+        if (userId) {
+          properties.user_id = userId;
+        }
+        data = orderEvent(properties);
+      } else {
+        if (userId) {
+          properties.user_id = userId;
+        }
```
Comment (shrouti1507) [inline integrations/Adroll/browser.js:86]: As we see,  
if (userId) {
          properties.user_id = userId;
        } 
is present in every if else block. Can we move it outside and implement once? It looks repeating.

File: integrations/Adroll/browser.js (hunk 8 above)
Comment (yashasvibajpai) [reply integrations/Adroll/browser.js]: Done.

File: integrations/Adroll/browser.js (hunk 9 above)
Comment (yashasvibajpai) [reply integrations/Adroll/browser.js]: Done. Thank you for pointing this

File: integrations/Adroll/browser.js [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+  }
+  // record_user fires the correct pixel in accordance with the event configured in the dashboard
+  // and the segment associated in adroll
+
+  page(rudderElement) {
+    logger.debug("=== In Adroll Page ===");
+    const { message } = rudderElement;
+
+    message.event = `Viewed ${message.name} ${message.category} Page`;
+    this.track(rudderElement);
```
Comment (utsabc) [inline integrations/Adroll/browser.js:94]: lets not call track
lets use 
```
window.__adroll.record_user, {
          adroll_segments: mapped_segment,
          name: name_of_event,
          path: window.location.pathname,
          referrer: window.document.referrer,
          search: window.location.search,
          title: window.document.title,
          url: window.location.href
        }
```

**Review by shrouti1507 — DISMISSED** [review]

**Review by utsabc — APPROVED** [review]

File: integrations/Adroll/browser.js [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+      pageFullName = `Viewed a Page`;
+    } else if (!message.name && message.category) {
+      pageFullName = `Viewed ${message.category} Page`;
+    } else if (message.name && !message.category) {
+      pageFullName = `Viewed ${message.name} Page`;
+    } else {
+      pageFullName = `Viewed ${message.name} ${message.category} Page`;
+    }
+    const segmentId = eventsHashmap[pageFullName.toLowerCase()];
+
```
Comment (ItsSudip) [inline integrations/Adroll/browser.js:104]: Check if the event name is there or not in the dashboard.

**Review by shrouti1507 — APPROVED** [review]
LGTM

**Review by ItsSudip — APPROVED** [review]

**Review by utsabc — APPROVED** [review]

**Pull request merged**